### ISSUE TITLE
[TESTNET BUG] Miner should mine blocks off of previous epoch (in most cases)

### ISSUE THREAD

**Issue by psq** [blockstack/stacks-blockchain#1760]
Let me preface this report with what I understand should be the ideal and desirable mining done by all miners.

> for each epoch, miners should do their best to mine a block that has for parent the block generated during the previous epoch

Which would lead to a nice linear blockchain with no forks.  Of course, this is real life, and with possibly faulty networking, faulty hardware, and even faulty code, it may happen that this won't always be the case, but this should be the exception (assuming every miner is trying to do its best and plays fair).

Now, back to reality, over the past 2+ weeks (with 3 resets so far, and 1000s of block, for which I was able to capture data), the only times when a block's parent is from the previous epoch are those were a miner wins the sortition 2 or more times in a row.

This would probably be the main reason why we see such a discrepancy between the number of burn blocks and the height of the stacks chain, https://github.com/blockstack/stacks-blockchain/issues/1745 would be an other one, but probably to a smaller extent, and I think there are more cases (which may be variations that could be addressed by fixing this and 1745).  The more miners mine, the more waste.

Included below are the full dump of week 27 run, week 28 run, and the aborted week 29 run (the second run would show the same thing).  If what I described as the ideal mining situation was happening, you'd see mostly `@@` (block's parent is the previous block mined, from a different miner) and sometimes `@+` (block's parent is the previous block mined, from the same miner) at the end of each line.  You will not see`@@`, not even once.  You see `@+` often, which seems to occur at an expected rate.

### chain analysis
[week 27](https://gist.github.com/psq/61923e67d990fc517b6c86f096350a08)
[week 28 - part 1](https://gist.github.com/psq/5a51114e353fd1af03744f79a3a3fa95)
[week 28 - part 2](https://gist.github.com/psq/e56e0f28748bfa5dec3c5988fc66f117)
[week 29 - run 1](https://gist.github.com/psq/b20e9a70e0030daf518fc0c4cbe7a3a2)

The analysis for each run uses a format similar to the one I used and described for https://github.com/blockstack/stacks-blockchain/issues/1745 with the addition of whether a block is based on the previous block from the same miner `@+` or from a different miner `@@`
The relevant bits:
- a `*` identifies the winning STX address (truncated to save space and increase readability)
- `p:` is the prefix for the parent's stacks block hash (also truncated)

for example, this bit is taken from week 27 and shows more than 1 miner mining consecutive blocks back to back, but only if they mined the previous block as well:
```
6852 ST2Z840ZWS* ST2TJRHDHM  3437 3437> 3437 br902 f65a4a0da9 p:675c225ed0 
6853 ST2Z840ZWS  ST3WCQ6S0D  ST2TJRHDHM* 3437 3437 3437 br906$ f146a91766 p:0f7f822cff 
6854 ST2Z840ZWS  ST2TJRHDHM* 3438 3438> 3438 br906$ e059430024 p:f146a91766 @+
6855 ST3WCQ6S0D  ST2Z840ZWS* ST2TJRHDHM  3438 3438 3438 br902 1c6e99634d p:f65a4a0da9 
6856 ST2Z840ZWS  ST2TJRHDHM* 3439 3439> 3439 br906$ 59b1999e95 p:e059430024 
6857 ST3WCQ6S0D* ST2Z840ZWS  ST2TJRHDHM  3439 3439 3439 br907 671b7b1835 p:e059430024 
6858 ST2Z840ZWS* ST2TJRHDHM  3439 3439 3439 br908 8b65802f19 p:e059430024 
6859 ST2Z840ZWS  ST3WCQ6S0D* ST2TJRHDHM  3440 3440> 3440 br906$ 68b3e52acc p:59b1999e95 
6860 ST2Z840ZWS* ST2TJRHDHM  3440 3440 3440 br908 1633d3cfb3 p:8b65802f19 
6861 ST2Z840ZWS* ST3WCQ6S0D  ST2TJRHDHM  3441 3441> 3441 br908 2e404b95c7 p:1633d3cfb3 @+
6862 ST2Z840ZWS  ST2TJRHDHM* 3441 3441 3441 br906$ 252595bfd1 p:68b3e52acc 
6863 ST3WCQ6S0D* ST2Z840ZWS  ST2TJRHDHM  3441 3441 3441 br909 a94a1d2aed p:68b3e52acc 
6864 ST2Z840ZWS* ST2TJRHDHM  3442 3442> 3442 br908 31ffdd3864 p:2e404b95c7 
6865 ST3WCQ6S0D  ST2Z840ZWS* ST2TJRHDHM  3443 3443> 3443 br908 3fe5b345e4 p:31ffdd3864 @+
6866 ST2Z840ZWS  ST2TJRHDHM* 3442 3442 3442 br906$ caf438d3db p:252595bfd1 
6867 ST3WCQ6S0D  ST2Z840ZWS  ST2TJRHDHM* 3443 3443 3443 br906$ 56c6d16871 p:caf438d3db @+
```
block mined based on burn block 6854 has the block mined for 6853 as parent, both mined by `ST2TJRHDHM` (go argon!)
but block mined based on burn block 6855 has a parent from the block mined for 6852, so this creates a fork
block mined based on burn block 6856 has a parent from the block mined for 6854, yet an other fork
...
next blocks mined from a previous block is at 6861, both mined by `ST2Z840ZWS` (that's my miner 😂)
then 6865, also from my miner
then 6867, from `ST2TJRHDHM` (that's argon again)

also worth noting are blocks 6856, 6857, and 6858, all with a parent mined from 6854, which may be a separate issue, but hard to tell for now.

5 consecutive blocks, mined by the same miner, 7009-7012:
```
7008 ST2Z840ZWS  ST2TJRHDHM* 3507 3507 3507 br934 0da03c0cad p:5d1f765f09 
7009 ST2Z840ZWS* ST2TJRHDHM  3509 3509> 3509 br915 da2694c33d p:88dd0f0f94 
7010 ST2Z840ZWS* ST2TJRHDHM  3510 3510> 3510 br915 836ff5d367 p:da2694c33d @+
7011 ST2Z840ZWS* ST2TJRHDHM  3511 3511> 3511 br915 fb127a9f96 p:836ff5d367 @+
7012 ST2Z840ZWS* ST2TJRHDHM  3512 3512> 3512 br915 9792538928 p:fb127a9f96 @+
7013 ST2Z840ZWS  ST2TJRHDHM* 3508 3508 3508 br923$ ae33908e01 p:01b801ab96 
```

**Comment by kantai** [comment]
Great analysis, @psq -- I've been visualizing the chain forks using this script (https://gist.github.com/kantai/86e8d062315050d90203d01f5976cda2#file-visualize_chain_forks-sh), and trying to get a sense for why we have some many short-lived forks as well. As you show, this is happening when consecutive blocks aren't mined by the same miner. To a certain extent, this is something that would be expected from short block times -- at a Bitcoin block time of 30 seconds, the winning miner needs to:

1. Discover the block from the Bitcoin node (this can take a couple of seconds)
2. Process the sortition, discover that it is the winner, and process its own tenure (maybe a couple seconds)
3. Push the block out to known nodes on the network
4. Have the block processed by other nodes _before_ they issue their own commitments

Network delays being what they are, 30 seconds isn't necessarily a ton of time to do all of the above, especially because a node wants to be sure to get it's commitment transaction into the Bitcoin mempool _before_ the next Bitcoin block.

However, the fact that miners _never_ successfully build off each other's blocks is probably indicative of something else going wrong, and I think that there's somewhat of a "bug" here. Our default configuration instructs miners to immediately mine a block after processing the previous sortition (this behavior is controlled via `wait_time_for_microblocks`, which defaults to 0). If, by default, every node waits for 0ms, then a node will almost _never_ have processed someone else's block before issuing a commitment.

So, I think what we need to do is change this default to something like 10-15s for Argon. Any thoughts, @lgalabru or @jcnelson ?

**Comment by psq** [comment]
@kantai impressive scripting-fu to generate `svg` with `bash`!

thank you for your thoughts, this helps.

I did some sampling (about 250 blocks) and here's what I found:
- burn blocks get received like clockwork with very little variation around the 30s mark, +/- 1s at most
- run tenure takes less than .5s, even when mining a block, and .2s on average, so not really an issue, but that's with very few transactions besides the coinbase.
- downloading a block if the node didn't win the sortition takes about 9.5s on average, with the fastest block taking 1.19s, but about 5% of the blocks may take 15s or more, and 16% of the blocks will not be available (more on this below).  It is probably fair to assume that blocks mined by argon would probably take less time than those submitting by other nodes.

For the above, tweaking `wait_time_for_microblocks` might help some, although if running a miner, I would rather skip a block than risk generating a block that will be rejected (the bitcoin fees have about quintupled over the past 2 weeks).  And there may be valid reasons for not getting a block (miner crashed, miner mined an invalid block, ...).  That said, if too many miners are too cautious, you could run the risk of not having anyone commit for a block.

Even though it was not possible from just the log I captured to analyze the block that did not show up during an epoch, while running a miner and running my analysis script, I could see pretty often that the miner was missing 5 or more blocks that eventually show up (I would think that would represent more than the 16% I measured in the case above), and that's probably a separate problem.  

My current thinking is that a better solution would be to:
- wait for the block before running `Run tenure`, the miner knows its hash from the winner of the sortition, with some heuristic to restart mining when a suitable block has been received (but this leaves the case where all miners stopped for the same reason)
- figure out why some blocks take a while to get propagated, even with a 30s block time, there should be enough time in most cases (on testnet/mainnet, this will be a lot longer, on average, but you can get very short block duration).  And it seems to work fine for burn blocks, so why would that not work for stacks blocks?

At least, that's what I can infer from just looking at what my miner sees, and I think I'm going to setup multiple nodes to go further...  For example any doc anywhere to setup my own argon master locally?

**Comment by psq** [comment]
> For example any doc anywhere to setup my own argon master locally

I managed to cobble something together that gives me a 2 node setup, including a regtest node, with blocks every 30s like argon, so hopefully I can understand better what is happening.  

While I was poking around trying to make my setup work, I suspect I may have found one reason why the stacks blocks don't always propagate as quickly as it would seem appropriate.  Not sure that's the main reason, but something like `download_interval` is set to `180` seconds, and will make the node throttle the downloads.  For 30 seconds blocks, this is probably not helping.

**Comment by jcnelson** [comment]
If you change the `BLOCK_DOWNLOAD_INTERVAL` in `src/net/downloads.rs` to something smaller, like 10, does the problem go away on your end?

**Comment by psq** [comment]
I actually tried a value of 10 last night, and even made it possible to set via the configuration file, and this seems to help on getting the blocks on a timely basis, I'm no longer seeing occurrences where it takes 3+ burn blocks to catch up on the stacks block. Of course, with all network going through the loopback, this is an ideal case and we'll need to test with remote nodes.
I'll also try with 5.

However, `run tenure` is still running too early on nodes that did not win the sortition, causing them to pick an older parent than they should, and I'm going to try a few things today to try to improve this, starting with the suggestion from @kantai to increase `wait_time_for_microblocks`, although I'm not convinced this is the *right* fix, but if this does not help, this would indicate other issues.  I'll update here when I have more data.

**Comment by psq** [comment]
quick update, as I did not spent much on this today, I was trying to understand an other issue with the miners from Phillip and Friedger, without much luck yet.  The miner says there are no UTXOs, but there are some, although lots with just 5500 (the dust value).

In any case, I did try with `wait_time_for_microblocks = 15000`, and `download_interval  = 10`.  It does seem very promising.  After 35 blocks, only 1 single fork with no abandoned offshoots.

```
222  0  -  - - -  
223 ST2Z840ZWS*(222) 1 1> 1 br1$ 4ff1840bfa p:0000000000 b:440857cb5b  
224 ST2Z840ZWS*(223) 2 2> 2 br1$ c8c6cd041f p:4ff1840bfa b:69e8ba5d5d @+ 
225 ST2Z840ZWS*(224) 3 3> 3 br1$ 818af19a84 p:c8c6cd041f b:2da7f63d1b @+ 
226 ST2Z840ZWS*(225) 4 4> 4 br1$ 3b3f8a3377 p:818af19a84 b:52b8cf67c1 @+ 
227 ST2CNRHV6T*(226) ST2Z840ZWS (226) 5 5> 5 br1$ 85a31d1bd0 p:3b3f8a3377 b:105ca75f9f @@ 
228 ST2CNRHV6T (227) ST2Z840ZWS*(227) 6 6> 6 br1$ 28d5443957 p:85a31d1bd0 b:7b9365cbf9 @@ 
229 ST2CNRHV6T (228) ST2Z840ZWS*(228) 7 7> 7 br1$ 72e16afa02 p:28d5443957 b:6e5d602ecd @+ 
230 ST2CNRHV6T (229) ST2Z840ZWS*(229) 8 8> 8 br1$ 5ba543e23b p:72e16afa02 b:72bc43675b @+ 
231 ST2CNRHV6T*(230) ST2Z840ZWS (230) 9 9> 9 br1$ e94510e8e0 p:5ba543e23b b:1127dfc884 @@ 
232 ST2CNRHV6T (231) ST2Z840ZWS*(231) 10 10> 10 br1$ 7075b074fb p:e94510e8e0 b:41f83fd2b2 @@ 
233 ST2CNRHV6T*(232) ST2Z840ZWS (232) 11 11> 11 br1$ 72d6857150 p:7075b074fb b:400df4a94e @@ 
234 ST2CNRHV6T (233) ST2Z840ZWS*(233) 12 12> 12 br1$ 2d656429b2 p:72d6857150 b:23bbf07bb9 @@ 
235 ST2CNRHV6T*(234) ST2Z840ZWS (234) 13 13> 13 br1$ 260fc79f55 p:2d656429b2 b:6f8ee770b3 @@ 
236 ST2CNRHV6T (235) ST2Z840ZWS*(235) 14 14> 14 br1$ 03adaa8f4d p:260fc79f55 b:11263940ef @@ 
237 ST2CNRHV6T*(236) ST2Z840ZWS (236) 15 15> 15 br1$ 501121e8b4 p:03adaa8f4d b:467c28c904 @@ 
238 ST2CNRHV6T (237) ST2Z840ZWS*(237) 16 16> 16 br1$ 1b656c123a p:501121e8b4 b:178483b153 @@ 
239 ST2CNRHV6T*(238) ST2Z840ZWS (238) 17 17> 17 br1$ 7075adb080 p:1b656c123a b:168ddbba77 @@ 
240 ST2CNRHV6T (239) ST2Z840ZWS*(239) 18 18> 18 br1$ 4a5efded5b p:7075adb080 b:1dc9a8050f @@ 
241 ST2CNRHV6T*(240) ST2Z840ZWS (240) 19 19> 19 br1$ 1fea57e748 p:4a5efded5b b:0b65d0ac8b @@ 
242 ST2CNRHV6T (241) ST2Z840ZWS*(241) 20 20> 20 br1$ 38dca5ccad p:1fea57e748 b:1aaba16125 @@ 
243 ST2CNRHV6T (242) ST2Z840ZWS*(242) 21 21> 21 br1$ b7cbfa72cb p:38dca5ccad b:0c6066774d @+ 
244 ST2CNRHV6T (243) ST2Z840ZWS*(243) 22 22> 22 br1$ a8b1b2f357 p:b7cbfa72cb b:03251c2e67 @+ 
245 ST2CNRHV6T (244) ST2Z840ZWS*(244) 23 23> 23 br1$ 45cff64e55 p:a8b1b2f357 b:1f703ec08e @+ 
246 ST2CNRHV6T*(245) ST2Z840ZWS (245) 24 24> 24 br1$ 4831d58197 p:45cff64e55 b:7af30635a8 @@ 
247 ST2CNRHV6T*(246) ST2Z840ZWS (246) 25 25> 25 br1$ a0d62b4059 p:4831d58197 b:42586886e5 @+ 
248 ST2CNRHV6T*(247) ST2Z840ZWS (247) 26 26> 26 br1$ ea04a26307 p:a0d62b4059 b:207ac2db46 @+ 
249 ST2CNRHV6T*(248) ST2Z840ZWS (248) 27 27> 27 br1$ 47a86a19e5 p:ea04a26307 b:1dc23dc568 @+ 
250 ST2CNRHV6T*(249) ST2Z840ZWS (249) 28 28> 28 br1$ d85062cd7d p:47a86a19e5 b:76e288a6a2 @+
251 ST2CNRHV6T (250) ST2Z840ZWS*(250) 29 29> 29 br1$ cb7235ac4a p:d85062cd7d b:6a42fa65f8 @@ 
252 ST2CNRHV6T*(251) ST2Z840ZWS (251) 30 30> 30 br1$ 463889dfed p:cb7235ac4a b:5d8c0f38c3 @@ 
253 ST2CNRHV6T (252) ST2Z840ZWS*(252) 31 31> 31 br1$ 119ac5a89b p:463889dfed b:1cabc7ca65 @@ 
254 ST2CNRHV6T*(253) ST2Z840ZWS (253) 32 32> 32 br1$ 83882df6b3 p:119ac5a89b b:154750df00 @@ 
255 ST2CNRHV6T*(254) ST2Z840ZWS (254) 33 33> 33 br1$ 4c319d3d80 p:83882df6b3 b:3ff81d9b64 @+ 
256 ST2CNRHV6T*(255) ST2Z840ZWS (255) 34 34> 34 br1$ c312061319 p:4c319d3d80 b:48121daa42 @+ 
257 ST2CNRHV6T*(256) ST2Z840ZWS (256) 35 35> 35 br1$ 8458da3356 p:c312061319 b:52262c1382 @+ 
```
`p:` is the hash of the parent block, `brxx` is the name of the fork, based on creation order (only one!), `@@` is when a different miner uses the previously mined block as its parent, `@+` when this is the same miner mining back to back, so in the above run, this looks great!

I did get both miners to panic at https://github.com/blockstack/stacks-blockchain/blob/master/testnet/stacks-node/src/neon_node.rs#L751-L753 shortly after starting, and also many bad blocks, but this could have been due to restarting the nodes after changing configuration parameters.  Hopefully more tomorrow.

**Comment by jcnelson** [comment]
> I actually tried a value of 10 last night, and even made it possible to set via the configuration file, and this seems to help on getting the blocks on a timely basis, I'm no longer seeing occurrences where it takes 3+ burn blocks to catch up on the stacks block. Of course, with all network going through the loopback, this is an ideal case and we'll need to test with remote nodes.
I'll also try with 5.

I'm glad to hear this at least worked!  Want to send a PR to `next` that makes this change?  Would want you to get credit for it in the commit history :)

**Comment by jcnelson** [comment]
> I did get both miners to panic at https://github.com/blockstack/stacks-blockchain/blob/master/testnet/stacks-node/src/neon_node.rs#L751-L753 shortly after starting, and also many bad blocks, but this could have been due to restarting the nodes after changing configuration parameters. Hopefully more tomorrow.

Yeah, there's no way to cleanly shut down the node right now.  If you restart it on existing chain state, you could be feeding it corrupt data, leading to a whole slew of different runtime panics.

**Comment by psq** [comment]
I'm glad to hear this at least worked! Want to send a PR to next that makes this change? Would want you to get credit for it in the commit history :) >

Will get something ready today, it'd be interesting to see how this behaves in the wild.

I let the setup run yesterday, and had the master stop at 2000 burn blocks.  Only 4 orphans blocks were generated over about 1800 burn blocks (my setup starts a bit earlier than argon), see the visualization using @kantai script https://gist.github.com/psq/1c15a78d779bc47cb4c7a4fad9cf6da3

**Comment by psq** [comment]
@jcnelson  created https://github.com/blockstack/stacks-blockchain/pull/1786

**Comment by timstackblock** [comment]
BUG TRACKING NOTE: Related to #1745 did not skip triage

**Comment by diwakergupta** [comment]
@psq can you verify if this is fixed by #1980 and close if so? Thanks!

**Comment by psq** [comment]
it looks like it is working much better, but there may be still some cases where this is not the case (https://github.com/blockstack/stacks-blockchain/pull/1980#issuecomment-712654207) so would like to keep it open for just a bit longer
